File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn- nothing? [s]
  (= "" (str/trim s)))

(defn- yelled? [s]
  (let [trimmed (str/trim s)]
    (and
     (re-find #"[A-Z]" trimmed)
     (= trimmed (str/upper-case trimmed)))))

(defn- question? [s]
  (->> s
       (str/trim)
       (last)
       (= \?)))

(defn- yelled-question? [s]
  (and (yelled? s) (question? s)))

(defn response-for
  "Takes some text and returns the response of Bob."
  [s]
  {:pre [(string? s)]
   :post [(string? %)]}
  (cond
    (yelled-question? s) "Calm down, I know what I'm doing!"
    (yelled? s) "Whoa, chill out!"
    (question? s) "Sure."
    (nothing? s) "Fine. Be that way!"
    :else "Whatever."))

File: javascript/bob/bob.js

export function hey(message) {

  const trimmedMessage = message.trim()
  const isEmpty = trimmedMessage === ''
  const hasLetter = /[A-Za-z]/.test(message)
  const isAllUppercase = message.toUpperCase() === message
  const isYelled = hasLetter & isAllUppercase
  const isQuestion = trimmedMessage.substr(trimmedMessage.length - 1) === '?'

  if (isEmpty)
    return 'Fine. Be that way!'

  if (isYelled && isQuestion)
    return 'Calm down, I know what I\'m doing!'

  if (isYelled)
    return 'Whoa, chill out!'

  if (isQuestion)
    return 'Sure.'

  return 'Whatever.'
}
